### PR TITLE
Pgta misc

### PR DESCRIPTION
**Pull request by jenswi-linaro**
Some small preparations for paged TAs

### CODE REVIEW

File: core/arch/arm/include/kernel/abort.h [hunk 1]
```diff
@@ -50,6 +50,8 @@ void abort_print_error(struct abort_info *ai);
 
 void abort_handler(uint32_t abort_type, struct thread_abort_regs *regs);
 
+bool abort_is_user_exception(struct abort_info *ai);
```
Comment (etiennec-linaro) [inline core/arch/arm/include/kernel/abort.h]: minor: abort_is_user_exception() is used only in abort.c. no need to export it.

File: core/arch/arm/mm/core_mmu.c [hunk 2]
```diff
@@ -959,6 +959,10 @@ static void check_pa_matches_va(void *va, paddr_t pa)
 		return;
 	}
 #ifdef CFG_WITH_PAGER
+	if (v >= CFG_TEE_LOAD_ADDR && v < core_mmu_linear_map_end) {
+		TEE_ASSERT(v == pa);
```
Comment (etiennec-linaro) [inline core/arch/arm/mm/core_mmu.c]: minor: why TEE_ASSERT() and not assert() ?
note: core_mmu.c uses assert() but in check_pa_matches_va() that uses TEE_ASSERT().

File: core/arch/arm/kernel/abort.c [hunk 3]
```diff
@@ -485,12 +485,14 @@ static bool is_vfp_fault(struct abort_info *ai __unused)
 
 static enum fault_type get_fault_type(struct abort_info *ai)
 {
-	if (is_user_exception(ai)) {
+	if (abort_is_user_exception(ai)) {
 		if (is_vfp_fault(ai))
 			return FAULT_TYPE_USER_TA_VFP;
+#ifndef CFG_WITH_PAGER
```
Comment (etiennec-linaro) [inline core/arch/arm/kernel/abort.c]: if pager is enable, ALIGNMENT aborts in user TA will panic TEE core in "case CORE_MMU_FAULT_ALIGNMENT" below. whereas old code simply trapped all user TA aborts for 'TA panic' handling.

File: core/arch/arm/include/kernel/abort.h (hunk 1 above)
Comment (jenswi-linaro) [inline core/arch/arm/include/kernel/abort.h]: I'm going to use it in a later patch

File: core/arch/arm/mm/core_mmu.c (hunk 2 above)
Comment (jenswi-linaro) [inline core/arch/arm/mm/core_mmu.c]: `TEE_ASSERT()` is unconditional, it will always evaluate and check the expression.
`assert()` is supposed to depend on `NDEBUG` and only evaluate and check expressions if `NDEBUG` isn't defined.

Our current `assert()` implementation doesn't honor `NDEBUG`, but may be changed into that some day.

I'm on the other hand tempted to remove `TEE_ASSERT()` and replace it with explicit `if (!expr) panic()`, that's something we would need to discuss first of course.

Comment (jforissier) [inline core/arch/arm/mm/core_mmu.c]: @jenswi-linaro +1 for fixing our `assert()` wrt. `NDEBUG` to properly follow the standard.

I also agree that `TEE_ASSERT` is confusing (is the expression always evaluated?), and I think your suggestion to simply remove it is good. While we're at it, I would suggest adding a reason string, like the Linux panic function: `if (!expr) panic("Error message")`. The implementation could print the message only, or the message plus the file/line info, depending on the debug level.

File: core/arch/arm/kernel/abort.c (hunk 3 above)
Comment (jenswi-linaro) [inline core/arch/arm/kernel/abort.c]: Thanks, I'll fix.

File: core/arch/arm/mm/core_mmu.c (hunk 2 above)
Comment (etiennec-linaro) [inline core/arch/arm/mm/core_mmu.c]: current assert already provides the function/line info where is asserted.
it not as explicit as a dedicated message of course.

Comment (jenswi-linaro) [inline core/arch/arm/mm/core_mmu.c]: Anyway, I think this is out of scope for this PR.

File: core/arch/arm/kernel/abort.c [hunk 4]
```diff
@@ -544,13 +550,16 @@ static enum fault_type get_fault_type(struct abort_info *ai)
...
+	bool handled;
 
 	set_abort_info(abort_type, regs, &ai);
 
 	switch (get_fault_type(&ai)) {
 	case FAULT_TYPE_IGNORE:
 		break;
 	case FAULT_TYPE_USER_TA_PANIC:
+		DMSG("[abort] abort in User mode (TA will panic)");
+		print_user_abort(&ai);
```
Comment (etiennec-linaro) [inline core/arch/arm/kernel/abort.c]: these 2 logs seems already provided in `get_fault_type()`

File: core/arch/arm/mm/core_mmu.c (hunk 2 above)
Comment (etiennec-linaro) [inline core/arch/arm/mm/core_mmu.c]: agree. discard my comment.
i note to have a look a assert() vs TEE_ASSERT(). they are redundant but regarding verbosity.

File: core/arch/arm/kernel/abort.c (hunk 4 above)
Comment (jenswi-linaro) [inline core/arch/arm/kernel/abort.c]: Sorry, I can't find that.

Comment (etiennec-linaro) [inline core/arch/arm/kernel/abort.c]: my mistake. - not +.

**Comment by etiennec-linaro** [comment]
`Reviewed-by: etienne carriere <etienne.carriere@linaro.org>`

**Comment by jenswi-linaro** [comment]
Tag applied
